File: src/systemd/lifecycle.rs
//! Service lifecycle: create / remove / start / stop / restart / enable / disable.

use std::collections::HashMap;
use std::fs;
use std::path::Path;

use crate::error::{Error, Result};
use crate::systemd::constants::SERVICE_PREFIX;
use crate::systemd::generation::generate_service_file;
use crate::systemd::manager::{validate_service_name, Manager};
use crate::systemd::mode::SystemdMode;
use crate::systemd::paths::{
    env_file_path, get_config_dir, get_log_dir, log_file_path, service_path, systemd_unit_name,
};
use crate::systemd::tasks::{task_env_file_path, task_log_file_path};
use crate::systemd::types::Service;

impl Manager {
    pub fn create(&self, svc: &Service) -> Result<()> {
        validate_service_name(&svc.name)?;
        if svc.mode != self.mode {
            return Err(Error::msg(format!(
                "service mode {:?} does not match manager mode {:?}",
                svc.mode, self.mode
            )));
        }
        if self.service_exists(&svc.name) {
            return Err(Error::ServiceAlreadyExists(svc.name.clone()));
        }

        let content = generate_service_file(svc)?;
        let path = service_path(&svc.name, svc.mode)?;
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        if matches!(svc.mode, SystemdMode::User) {
            if let Ok(log_dir) = get_log_dir(svc.mode) {
                let _ = fs::create_dir_all(log_dir);
            }
        }

        fs::write(&path, content)?;

        if !svc.environment.is_empty() {
            self.create_env_file(&svc.name, &svc.environment)
                .map_err(|e| Error::msg(format!("create environment file: {e}")))?;
        }

        self.daemon_reload()?;
        Ok(())
    }

    pub fn create_env_file(&self, service_name: &str, env: &HashMap<String, String>) -> Result<()> {
        let env_dir = get_config_dir(self.mode)?;
        fs::create_dir_all(&env_dir)?;
        let mut content = String::new();
        content.push_str(&format!(
            "# Environment variables for smdctl service: {service_name}\n"
        ));
        content.push_str(&format!(
            "# Edit this file with: smdctl env {service_name}\n\n"
        ));
        for (k, v) in env {
            content.push_str(&format!("{k}={v}\n"));
        }
        let env_path = env_dir.join(format!("{service_name}.env"));
        fs::write(env_path, content)?;
        Ok(())
    }

    pub fn service_exists(&self, name: &str) -> bool {
        if service_path(name, self.mode)
            .map(|p| p.exists())
            .unwrap_or(false)
        {
            return true;
        }
        let other = match self.mode {
            SystemdMode::User => SystemdMode::System,
            SystemdMode::System => SystemdMode::User,
        };
        service_path(name, other)
            .map(|p| p.exists())
            .unwrap_or(false)
    }

    pub fn start(&self, name: &str) -> Result<()> {
        if !self.service_exists(name) {
            return Err(Error::ServiceNotFound(name.to_string()));
        }
        let unit = systemd_unit_name(name);
        self.start_unit(&unit)
    }

    pub fn stop(&self, name: &str) -> Result<()> {
        if !self.service_exists(name) {
            return Err(Error::ServiceNotFound(name.to_string()));
        }
        let unit = systemd_unit_name(name);
        self.stop_unit(&unit)
    }

    pub fn restart(&self, name: &str) -> Result<()> {
        if !self.service_exists(name) {
            return Err(Error::ServiceNotFound(name.to_string()));
        }
        let unit = systemd_unit_name(name);
        self.exec.run(self.mode, &["restart", &unit])
    }

    pub fn enable(&self, name: &str) -> Result<()> {
        if !self.service_exists(name) {
            return Err(Error::ServiceNotFound(name.to_string()));
        }
        let unit = systemd_unit_name(name);
        self.enable_unit(&unit)
    }

    pub fn disable(&self, name: &str) -> Result<()> {
        if !self.service_exists(name) {
            return Err(Error::ServiceNotFound(name.to_string()));
        }
        let unit = systemd_unit_name(name);
        self.disable_unit(&unit)
    }

    pub fn remove(&self, name: &str) -> Result<()> {
        if !self.service_exists(name) {
            return Err(Error::ServiceNotFound(name.to_string()));
        }
        let _ = self.remove_tasks(name);
        let _ = self.stop(name);
        let _ = self.disable(name);

        let path = service_path(name, self.mode)?;
        fs::remove_file(path)?;

        if let Ok(envp) = env_file_path(name, self.mode) {
            let _ = fs::remove_file(envp);
        }

        if let Some(lp) = log_file_path(name, self.mode) {
            let _ = fs::remove_file(lp);
        }

        self.daemon_reload()?;
        Ok(())
    }

    pub fn remove_tasks(&self, service_name: &str) -> Result<()> {
        let svc_path = service_path(service_name, self.mode)?;
        let unit_dir = svc_path.parent().map(Path::to_path_buf).unwrap_or_default();
        let prefix = format!("{SERVICE_PREFIX}{service_name}-task-");

        let entries = match fs::read_dir(&unit_dir) {
            Ok(e) => e,
            Err(_) => return Ok(()),
        };

        for entry in entries.flatten() {
            let path = entry.path();
            let fname = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
            if !fname.ends_with(".timer") {
                continue;
            }
            if !fname.starts_with(&prefix) {
                continue;
            }

            let base = fname.trim_end_matches(".timer");
            let task_name = base.strip_prefix(&prefix).unwrap_or("");

            let _ = self.stop_unit(fname);
            let _ = self.disable_unit(fname);
            let _ = fs::remove_file(&path);

            let one_shot = format!("{base}.service");
            let _ = self.stop_unit(&one_shot);
            let _ = fs::remove_file(unit_dir.join(&one_shot));

            if !task_name.is_empty() {
                if let Ok(tep) = task_env_file_path(service_name, task_name, self.mode) {
                    let _ = fs::remove_file(tep);
                }
                if let Some(tlp) = task_log_file_path(service_name, task_name, self.mode) {
                    let _ = fs::remove_file(tlp);
                }
            }
        }

        Ok(())
    }

    pub fn daemon_reload(&self) -> Result<()> {
        self.exec.run(self.mode, &["daemon-reload"])
    }

    pub(super) fn start_unit(&self, unit: &str) -> Result<()> {
        self.exec.run(self.mode, &["start", unit])
    }

    pub(super) fn stop_unit(&self, unit: &str) -> Result<()> {
        self.exec.run(self.mode, &["stop", unit])
    }

    pub(super) fn enable_unit(&self, unit: &str) -> Result<()> {
        self.exec.run(self.mode, &["enable", unit])
    }

    pub(super) fn disable_unit(&self, unit: &str) -> Result<()> {
        self.exec.run(self.mode, &["disable", unit])
    }
}

File: internal/systemd/lifecycle.go
package systemd

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

func (m *Manager) DaemonReload() error           { return m.exec().Run(m.Mode, "daemon-reload") }
func (m *Manager) startUnit(unit string) error   { return m.exec().Run(m.Mode, "start", unit) }
func (m *Manager) stopUnit(unit string) error    { return m.exec().Run(m.Mode, "stop", unit) }
func (m *Manager) enableUnit(unit string) error  { return m.exec().Run(m.Mode, "enable", unit) }
func (m *Manager) disableUnit(unit string) error { return m.exec().Run(m.Mode, "disable", unit) }

func (m *Manager) Create(svc *Service) error {
	if err := ValidateServiceName(svc.Name); err != nil {
		return err
	}
	if svc.Mode != m.Mode {
		return fmt.Errorf("service mode %s does not match manager mode %s", modeDebug(svc.Mode), modeDebug(m.Mode))
	}
	if m.ServiceExists(svc.Name) {
		return fmt.Errorf("service already exists: %s", svc.Name)
	}
	content, err := GenerateServiceFile(svc)
	if err != nil {
		return err
	}
	path, err := ServicePath(svc.Name, svc.Mode)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0777); err != nil {
		return err
	}
	if svc.Mode == User {
		if dir, err := LogDir(svc.Mode); err == nil {
			_ = os.MkdirAll(dir, 0777)
		}
	}
	if err := os.WriteFile(path, []byte(content), 0666); err != nil {
		return err
	}
	if len(svc.Environment) > 0 {
		if err := m.CreateEnvFile(svc.Name, svc.Environment); err != nil {
			return fmt.Errorf("create environment file: %w", err)
		}
	}
	return m.DaemonReload()
}
func (m *Manager) CreateEnvFile(name string, env map[string]string) error {
	dir, err := ConfigDir(m.Mode)
	if err != nil {
		return err
	}
	if err = os.MkdirAll(dir, 0777); err != nil {
		return err
	}
	var b strings.Builder
	fmt.Fprintf(&b, "# Environment variables for smdctl service: %s\n# Edit this file with: smdctl env %s\n\n", name, name)
	for k, v := range env {
		fmt.Fprintf(&b, "%s=%s\n", k, v)
	}
	return os.WriteFile(filepath.Join(dir, name+".env"), []byte(b.String()), 0666)
}
func (m *Manager) Start(name string) error {
	if !m.ServiceExists(name) {
		return fmt.Errorf("service not found: %s", name)
	}
	return m.startUnit(UnitName(name))
}
func (m *Manager) Stop(name string) error {
	if !m.ServiceExists(name) {
		return fmt.Errorf("service not found: %s", name)
	}
	return m.stopUnit(UnitName(name))
}
func (m *Manager) Restart(name string) error {
	if !m.ServiceExists(name) {
		return fmt.Errorf("service not found: %s", name)
	}
	return m.exec().Run(m.Mode, "restart", UnitName(name))
}
func (m *Manager) Enable(name string) error {
	if !m.ServiceExists(name) {
		return fmt.Errorf("service not found: %s", name)
	}
	return m.enableUnit(UnitName(name))
}
func (m *Manager) Disable(name string) error {
	if !m.ServiceExists(name) {
		return fmt.Errorf("service not found: %s", name)
	}
	return m.disableUnit(UnitName(name))
}
func (m *Manager) Remove(name string) error {
	if !m.ServiceExists(name) {
		return fmt.Errorf("service not found: %s", name)
	}
	_ = m.RemoveTasks(name)
	_ = m.Stop(name)
	_ = m.Disable(name)
	path, err := ServicePath(name, m.Mode)
	if err != nil {
		return err
	}
	if err = os.Remove(path); err != nil {
		return err
	}
	if p, e := EnvFilePath(name, m.Mode); e == nil {
		_ = os.Remove(p)
	}
	if p := LogFilePath(name, m.Mode); p != "" {
		_ = os.Remove(p)
	}
	return m.DaemonReload()
}
func (m *Manager) RemoveTasks(serviceName string) error {
	svcPath, err := ServicePath(serviceName, m.Mode)
	if err != nil {
		return err
	}
	unitDir := filepath.Dir(svcPath)
	prefix := fmt.Sprintf("%s%s-task-", ServicePrefix, serviceName)
	entries, err := os.ReadDir(unitDir)
	if err != nil {
		return nil
	}
	for _, entry := range entries {
		name := entry.Name()
		if !strings.HasPrefix(name, prefix) || !strings.HasSuffix(name, ".timer") {
			continue
		}
		base := strings.TrimSuffix(name, ".timer")
		taskName := strings.TrimPrefix(base, prefix)
		_ = m.stopUnit(name)
		_ = m.disableUnit(name)
		_ = os.Remove(filepath.Join(unitDir, name))
		oneShot := base + ".service"
		_ = m.stopUnit(oneShot)
		_ = os.Remove(filepath.Join(unitDir, oneShot))
		if taskName != "" {
			if p, e := TaskEnvFilePath(serviceName, taskName, m.Mode); e == nil {
				_ = os.Remove(p)
			}
			if p := TaskLogFilePath(serviceName, taskName, m.Mode); p != "" {
				_ = os.Remove(p)
			}
		}
	}
	return nil
}
